File: src/turbopuffer_fs/checks.py
"""Pure precondition checks for mutation plans."""

from __future__ import annotations

import errno


def rows_for(results: dict[str, dict[str, object]], name: str) -> list[dict[str, object]]:
    return list(results.get(name, {}).get("rows", []))


def first_row(results: dict[str, dict[str, object]], name: str) -> dict[str, object] | None:
    rows = rows_for(results, name)
    return rows[0] if rows else None


def _rows_by_path(rows: list[dict[str, object]]) -> dict[str, dict[str, object]]:
    return {str(row["path"]): row for row in rows}


def check_mkdir_preconditions(context: dict[str, object], results: dict[str, dict[str, object]]) -> None:
    path = str(context["path"])
    existing = _rows_by_path(rows_for(results, "existing"))
    parent_paths = list(context["parent_paths"])
    for ancestor in parent_paths:
        row = existing.get(ancestor)
        if row is not None and row.get("kind") != "dir":
            raise NotADirectoryError(ancestor)
    target = existing.get(path)
    if target is not None and target.get("kind") != "dir":
        raise FileExistsError(path)


def check_put_preconditions(context: dict[str, object], results: dict[str, dict[str, object]]) -> None:
    path = str(context["path"])
    existing = _rows_by_path(rows_for(results, "existing"))
    parent_paths = list(context["parent_paths"])
    for ancestor in parent_paths:
        row = existing.get(ancestor)
        if row is not None and row.get("kind") != "dir":
            raise NotADirectoryError(ancestor)
    target = existing.get(path)
    if target is not None and target.get("kind") == "dir":
        raise IsADirectoryError(path)


def check_rm_preconditions(context: dict[str, object], results: dict[str, dict[str, object]]) -> None:
    path = str(context["path"])
    recursive = bool(context.get("recursive", False))
    target = first_row(results, "target")
    if target is None:
        return
    if recursive:
        return
    if target.get("kind") == "dir" and rows_for(results, "child_probe"):
        raise OSError(errno.ENOTEMPTY, "Directory not empty", path)


CHECKS = {
    "mkdir_preconditions": check_mkdir_preconditions,
    "put_preconditions": check_put_preconditions,
    "rm_preconditions": check_rm_preconditions,
}


def run_check(name: str, context: dict[str, object], results: dict[str, dict[str, object]]) -> None:
    CHECKS[name](context, results)

File: packages/turbopuffer-fs/src/checks.ts
import type { AnyObject, ExecuteResults, RowLike } from "./types.js";

const rowsFor = (results: ExecuteResults, name: string): RowLike[] =>
  ((results[name] as Record<string, unknown> | undefined)?.rows as RowLike[] | undefined) ?? [];

const firstRow = (results: ExecuteResults, name: string): RowLike | null =>
  rowsFor(results, name)[0] ?? null;

const rowsByPath = (rows: RowLike[]): Record<string, RowLike> =>
  Object.fromEntries(rows.map((row) => [String(row.path), row]));

function checkMkdirPreconditions(
  context: Record<string, unknown>,
  results: ExecuteResults,
): void {
  const path = String(context.path);
  const existing = rowsByPath(rowsFor(results, "existing"));
  const parentPaths = (context.parentPaths as string[] | undefined) ?? [];
  for (const ancestor of parentPaths) {
    const row = existing[ancestor];
    if (row && row.kind !== "dir") {
      throw new Error(`NotADirectoryError:${ancestor}`);
    }
  }
  const target = existing[path];
  if (target && target.kind !== "dir") {
    throw new Error(`FileExistsError:${path}`);
  }
}

function checkPutPreconditions(
  context: Record<string, unknown>,
  results: ExecuteResults,
): void {
  const path = String(context.path);
  const existing = rowsByPath(rowsFor(results, "existing"));
  const parentPaths = (context.parentPaths as string[] | undefined) ?? [];
  for (const ancestor of parentPaths) {
    const row = existing[ancestor];
    if (row && row.kind !== "dir") {
      throw new Error(`NotADirectoryError:${ancestor}`);
    }
  }
  const target = existing[path];
  if (target && target.kind === "dir") {
    throw new Error(`IsADirectoryError:${path}`);
  }
}

function checkRmPreconditions(
  context: Record<string, unknown>,
  results: ExecuteResults,
): void {
  const path = String(context.path);
  const recursive = Boolean(context.recursive);
  const target = firstRow(results, "target");
  if (!target || recursive) {
    return;
  }
  if (target.kind === "dir" && rowsFor(results, "child_probe").length > 0) {
    throw new Error(`DirectoryNotEmptyError:${path}`);
  }
}

const CHECKS = {
  mkdir_preconditions: checkMkdirPreconditions,
  put_preconditions: checkPutPreconditions,
  rm_preconditions: checkRmPreconditions,
} as const;

export function runCheck(
  name: string,
  context: Record<string, unknown>,
  results: ExecuteResults,
): void {
  const check = CHECKS[name as keyof typeof CHECKS];
  if (!check) {
    throw new Error(`UnknownCheck:${name}`);
  }
  check(context, results);
}
